src/wallet/multi_tracker.rs

```
@@ -0,0 +1,405 @@
+}
+
+/// Represents a collection of owned and spendable `UTXO`s, `ExtendedDescriptor`s and associated
+/// `Transaction`s.
+pub trait MultiTracker: MultiTrackerInner {
+    /// Obtains parent [`ExtendedDescriptor`] and child index of the provided `ScriptPubKey`.
+    /// Note that if the script is not owned, None shoud be returned.
+    fn get_path_of_script_pubkey(
+        &self,
+        script: &Script,
+    ) -> Result<Option<(DescriptorItem, u32)>, Error> {
+        let descriptors = self.iter_descriptors()?.collect::<Vec<_>>();
+
+        let secp = SecpCtx::new();
+
+        for index in 0..2100_u32 {
+            for item in &descriptors {
+                let desc = item.descriptor.clone();
+                let derived_script = desc.as_derived(index, &secp).script_pubkey();
+                if script == &derived_script {
+                    return Ok(Some((item.clone(), index)));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Obtain a local UTXO given the provided outpoint.
+    /// The default implementation is very inefficient and should be overloaded.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Result<Option<LocalUtxo>, Error> {
+        Ok(self.iter_unspent()?.find_map(|(utxo, _)| {
+            if &utxo.outpoint == outpoint {
+                Some(utxo)
+            } else {
+                None
+            }
+        }))
+    }
+
+    /// Obtain output of the provided outpoint.
+    /// Output may not be owned by us, just be part of a transaction that we are keeping track of.
+    fn get_output(&self, outpoint: &OutPoint) -> Result<Option<TxOut>, Error> {
+        self.get_tx(&outpoint.txid)?.map_or_else(
+            || Err(Error::InvalidOutpoint(*outpoint)),
+            |tx_item| {
+                Ok(tx_item
+                    .raw
+                    .output
+                    .get(outpoint.vout as usize)
+                    .map(Clone::clone))
```

Should we return an error if a tx doesn't have the requested vout? I kinda feel like these are swapped around, we should return `None` if we don't know about the outpoint, but if we know the tx and the vout is missing, then it's clearly invalid

---

Resolved in 7649de62edc4610107c852185e87780a05a92ee6